### ISSUE TITLE
SCS Harbor failing to push images

### ISSUE THREAD

**Issue by matofeder** [SovereignCloudStack/k8s-harbor#62]
We've identified instances of failures when pushing to the Harbor registry. This problem occurs consistently when pushing large images (>1GB), but can also occur randomly with smaller images (~100MB)

Docker push is retrying, e.g.:
 ```
 $ docker push registry.scs.community/test-matej/imagetest1g:latest 
The push refers to repository [registry.scs.community/test-matej/imagetest1g]
b557b243c90e: Retrying in 8 seconds 
d101c9453715: Layer already exists 
```
and then failed with ` write: connection reset by peer`:

```
$ docker push registry.scs.community/test-matej/imagetest1g:latest 
The push refers to repository [registry.scs.community/test-matej/imagetest1g]
b557b243c90e: Pushing [==================================================>]  1.074GB/1.074GB
d101c9453715: Layer already exists 
write tcp 192.168.88.247:40116->213.131.230.28:443: write: connection reset by peer
```

K8s Ingres nginx controller logs HTTP400:
```
 84.245.97.28 - - [04/Mar/2024:07:42:28 +0000] "PATCH /v2/test-matej/imagetest1g/blobs/uploads/717a9b86-e83d-4457-9f4e-cf890ae60bd2?_state=t0uBz4GwYHVE_Q0q-uE51RjGLpc9HTD0w5709LOgg2h7Ik5hbWUiOiJ0ZXN0LW1hdGVq │
│ L2ltYWdldGVzdDFnIiwiVVVJRCI6IjcxN2E5Yjg2LWU4M2QtNDQ1Ny05ZjRlLWNmODkwYWU2MGJkMiIsIk9mZnNldCI6MCwiU3RhcnRlZEF0IjoiMjAyNC0wMy0wNFQwNzozNTo0MS4zODg0NTgzMloifQ%3D%3D HTTP/1.1" 400 0 "-" "docker/25.0.3 go/go1.21. │
│ 6 git-commit/f417435 kernel/6.5.0-21-generic os/linux arch/amd64 UpstreamClient(Docker-Client/25.0.3 \x5C(linux\x5C))" 987595584 406.505 [default-harbor-harbor-core-80] [] 192.168.42.185:8080 0 406.505 - 0b │
│ d8ccfedb6048149912b0c81990f651  
```

Harbor Core logs the proxy error:
```
2024/03/04 07:36:15 http: proxy error: context canceled                                                                                                                                                       
2024/03/04 07:42:58 http: proxy error: context canceled 
```

Harbor version: v2.9.2 (deployed in the K8s cluster, via Helm)
K8s version:  v1.27.5 
Ingress controller version: v1.8.1

**Comment by matofeder** [comment]
We suspected the ingress proxy settings, so we tried the recommended setting from [here](https://github.com/goharbor/harbor/issues/16623) and [here](https://docs.daocloud.io/en/kangaroo/best-practice/harbor-nginx.html#scenario-2-failed-helm-repo-operations)

tl;dr; we tried to increase timeouts (proxy_send_timeout 900; proxy_read_timeout 900;) and turn off proxy buffering (proxy-request-buffering: off; proxy-buffering: off).

Unfortunately, the above did not help.

The next step could be, to try to upgrade the Harbor and/or Nginx ingress controller.

**Comment by matofeder** [comment]
Note: Harbor Upgrade failed, we have to wait until the Harbor Helm chart is fixed. https://github.com/SovereignCloudStack/k8s-harbor/issues/63

**Comment by chess-knight** [comment]
Ingress controller version upgraded to v1.9.6

**Comment by chess-knight** [comment]
After upgrades #64, #65 and `proxy-request-buffering: off` in #67,  I was able to successfully push images ~ 5GB.

@matofeder @jschoone can you confirm that it works for you?

**Comment by matofeder** [comment]
I've tested the case: push a large (>6GB) image to the SCS harbor as follows:

```bash
# Dockerfile that just creates 6G file
$ cat LargeDockerfile
FROM ubuntu:latest
RUN dd if=/dev/urandom of=sample.txt bs=1G count=6 iflag=fullblock

# Build image
$ docker build -f LargeDockerfile -t registry.scs.community/test-matej/6G:test .

# Check size
$ docker image ls registry.scs.community/test-matej/6g
REPOSITORY                             TAG       IMAGE ID       CREATED          SIZE
registry.scs.community/test-matej/6g   test      e2e38690d559   15 minutes ago   6.52GB

# Push image to the SCS Harbor registry
$ docker push registry.scs.community/test-matej/6g:test
The push refers to repository [registry.scs.community/test-matej/6g]
f19e5b3ec73a: Pushing [==================================================>]  6.442GB
e0a9f5911802: Pushed 
f19e5b3ec73a: Pushed 
test: digest: sha256:e1179059c5d34e7b9058be1ebb63e9e484ad54e7f44a7566ffc1014ea9217bfb size: 743

```

So based on the above I can confirm that I was able to push the 6GB+ container image to the SCS Harbor registry.
